### DISCUSSION TITLE
What is the parameter of optimize_acqf

### DISCUSSION THREAD

**Discussion by makinzm** [meta-pytorch/botorch#1068] [Q&A] [answered]
# 1. What I would like to do

- I would like to understand what `optimize_acqf` does, especially `num_restarts` , `raw_samples` and the flow.
- Now, I think I may understand the rough abstract but don't have confidence. I would appreciate it, If you could teach me.

# 2. What I did

- I read this paper roughly :
    -  Balandat, Maximilian, et al. "BoTorch: A Framework for Efficient Monte-Carlo Bayesian Optimization." arXiv preprint arXiv:1910.06403 (2019). (https://arxiv.org/abs/1910.06403)
- I read this issue :
    - What is the role of the 'raw samplers' in optimize_acqf? #366
    - (https://github.com/pytorch/botorch/issues/366)

# 3. How I understand now

If I use the following code :

```python
candidate, _ = optimize_acqf(
    aq, bounds=bounds, q=q,
    num_restarts=num_restarts,
    raw_samples=raw_samples,
)
```

I think the flow is the following roughtly:

1. Sample $N$(`=raw_samples`) (in `bounds`) $\xi_i\sim f(x^q|\mathcal{D})$ by fixing the random variable by SSA and using MCMC.

2. Compute the acquisition function value $\hat{\alpha}(\xi)=\frac{1}{N}\sum\alpha(\xi_i)$ by using $N$ samples.

3. SGD Starts point is $\xi$ and the loop $N'$(`=num_restarts`) SGD starts ($\xi_{i,j+1}=\xi_{i,j}-\frac{1}{N}\sum\nabla \alpha(\xi_{i,j})$).

4. get $\xi$ and $\hat{\alpha}(\xi)$

I think it is not correct , but I cannot what is right.

I would appreciate it, If you could teach me.

Thank you for reading.

**Comment by Balandat** [comment, marked as answer]
So the flow is
1. Sample (using QMC) `raw_samples` points `X`  in `bounds`.
2. Compute the acquisition function value `Z := alpha(X)` for each of these samples. 
3. Apply a Boltzmann sampling heuristic ([this one](https://github.com/pytorch/botorch/blob/3318c06333166328c95d24edd1055d4ef4bded70/botorch/optim/initializers.py#L458)) that selects `num_restarts ` points from `X` (without replacement) with probability proportional to `exp(eta * Z)`, where `Z = (Y - mean(Y)) / std(Y)` and `eta` is a temperature parameter.
4. Use the points that were picked by the heuristic as initial conditions for multi-start L-BFGS-B optimization.

**Reply by makinzm** [reply]
Thank you for teaching me.
I will re-think this flow.
